README.md
# NAME

Search::Typesense - Perl interface to Typesense search engine.

# SYNOPSIS

    my $typesense = Search::Typesense->new(
        host      => $host,    # required
        api_key   => $key,     # required
        port      => $port,    # defaults to 8108
        use_https => $bool,    # defaults to true
    );
    
    my $results = $typesense->search(
        $collection_name,
        { q => 'Search String' },
    );
    if ( $results->{found} ) {
        foreach my $hit ( @{ $results->{hits} } ) {
            ...;
        }
    }

[Check here for a comparison to ElasticSearch and similar technologies](https://typesense.org/typesense-vs-algolia-vs-elasticsearch-vs-meilisearch/).

# DESCRIPTION

**ALPHA CODE**. The interface can and will change without warning.

This is an interface to the [Typesense](https://typesense.org/) search
engine. Most methods will do one of three things:

- Return results as defined in the Typesense documentation (listed per section)
- Return nothing if Typesense returns a 404.
- `croak` if Typesense returns an error.

# CONSTRUCTOR

The constructor takes a list (or hashref) of key/value pairs.

    my $typesense = Search::Typesense->new(
        host      => $host,    # required
        api_key   => $key,     # required
        port      => $port,    # defaults to 8108
        use_https => $bool,    # defaults to true
    );

## `api_key`

The api key to which will be sent as the `X-TYPESENSE-API-KEY` header.

## `host`

The hostname to connect to.

## `port`

Optional port number to connect to. Defaults to 8108 if not supplied.

## `use_https`

Optional boolean. Whether or not to connect to Typesense over https. Default true.

# METHODS

## `collections`

    my $collections = $typesense->collections;
    my $collection  = $collections->get($collection_name);
    my $results     = $collections->search($collection_name, {q => 'London'});

Returns an instance of [Search::Typesense::Collection](https://metacpan.org/pod/Search::Typesense::Collection) for managing Typesense collections.

## `search`

    my $results = $typesense->search($collection_name, {q => 'London'});

Shorthand that delegated to `$typesense->collections->search(...)`.

We do this hear mainly because this is the common case.

## `documents`

    my $documents = $typesense->documents;
    my $document  = $documents->delete($collection_name, $document_id);

Returns an instance of [Search::Typesense::Document](https://metacpan.org/pod/Search::Typesense::Document) for managing Typesense documents.

## `assert_is_running`

    $typesense->assert_is_running;

This does nothing if we can connect to Typesense. Otherwise, this method will
`croak` with a message explaining the error.

## `typesense_version`

    my $version = $typesense->typesense_version;

Returns an instance of [Search::Typesense::Version](https://metacpan.org/pod/Search::Typesense::Version).

If your version of Typesense is older than `0.8.0`, this method will return
nothing.

# AUTHOR

Curtis "Ovid" Poe, `<ovid at allaroundtheworld.fr>`

# BUGS

Please report any bugs or feature requests to
`https://github.com/Ovid/Search-Typesense/issues`.  I will be notified, and
then you'll automatically be notified of progress on your bug as I make
changes.

# SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Search::Typesense

You can also look for information at:

- Github Repo

    [https://github.com/Ovid/Search-Typesense/](https://github.com/Ovid/Search-Typesense/)

- Issue Tracker

    [https://github.com/Ovid/Search-Typesense/issues](https://github.com/Ovid/Search-Typesense/issues)

- Search CPAN

    [https://metacpan.org/release/Search-Typesense](https://metacpan.org/release/Search-Typesense)

# ACKNOWLEDGEMENTS

Thanks for Sebastian Reidel and Matt Trout for feedback.

# LICENSE AND COPYRIGHT

This software is Copyright (c) 2021 by Curtis "Ovid" Poe.

This is free software, licensed under:

    The Artistic License 2.0 (GPL Compatible)


Changes
{{$NEXT}}

0.08      2021-03-31 09:07:28 CEST

          - Bugfix in tests: if Typesense is not running, don't SKIP tests
            just because we don't have a server. We currently cache test
            results.

0.07      2021-03-26 13:39:29 CET

          - Tests can now cache their output and be rerun without a Typesense
            server. Delete the t/cache directory to require a Typesense server
            again.
          - Consolidate the http request behavior into a single role,
            Search::Typesense::Role::Request.
          - Standardized HTTP request args in _GET, _POST, _PATCH, and _DELETE
            methods.
          - Added Perl::Critic and Perl::Tidy tests to xt/

0.06      2021-03-23 18:09:49 CET
          - No significant (non-doc) changes. The version number was bumped
            automatically by Dist::Zilla as I was trying to force a new
            release.

0.05      2021-03-23 17:59:07 CET
          - Collection and Document behaviors are now moved into their own
            subclasses.

0.04      2021-03-23 17:01:20 CET
          - Move author tests to xt directory.
          - Expose the Typesense version (this appears to not be documented)

0.03      2021-03-23 09:32:02 CET
          - Fix dependencies in the dist.ini file. This should resolve
            https://github.com/Ovid/Search-Typesense/issues/1
          - Start a proper Test::Search::Typesense module. It's only a stub
            right now, but it simplifies some common code.
          - Lay the groundwork for more i18n tests in the future.

0.02      2021-03-22 15:38:24 CET
          - port argument now defaults to 8108
          - update_document() was broken. It's now fixed and tested.
          - Renamed purge to delete_all_collections()
          - Bump hard dependency on Perl to 5.16.0
          - Make the Moo code more idiomatic
          - Assert that the port is a PositiveInt
          - Add kraih and mst to acknowledgements

0.01      2021-03-22 13:22:28 CET
          - Basic interface to Typesense.



t/lib/Test/Search/Typesense.pm
package Test::Search::Typesense;

use Moo;
use Test::Most ();
use Test::Search::Typesense::Cached;
use Search::Typesense::Types qw(
  InstanceOf
);

has typesense => (
    is      => 'ro',
    isa     => InstanceOf ['Search::Typesense'],
    builder => '_build_typesense',
);

sub _build_typesense {
    my $self      = shift;
    my $typesense;
    eval {
        $typesense = Test::Search::Typesense::Cached->new(
            use_https => 0,
            host      => 'localhost',
            port      => 7777,
            api_key   => 777,
        );
    };
    if ($typesense) {
        $typesense->collections->delete_all;
        return $typesense;
    }

    my $reason = $@;
    Test::Most::explain(<<"END");
If they don't have Typesense running, we skip the tests and give them the
information they need to get the tests running. However, if they're running a
bizarrely old version of Typesense (< 0.8.0), we don't guarantee support and
we bail out.

Error reason: $reason
END
    Test::More::plan( skip_all =>
          "Typesense does not appear to be running. See the CONTRIBUTING.md document with this distribution."
    );
    unless ( $typesense->typesense_version ) {
        Test::More::diag(
            "https://github.com/typesense/typesense-api-spec/commit/778ad3e0d2bdf23e6ccc1b23113ae6f48ec345fb"
        );
        Test::More::BAIL_OUT(
            "You're using a version of Typesense earlier than 0.8.0.");
    }
}

# If, for some strange reason, we've still hit an existing Typesense database,
# minimize the chance of hitting a valid collection
sub company_collection_name {
    'company_XXX_this_will_be_deleted_after_testing_XXX';
}

sub company_collection_definition {
    my $self = shift;
    return {
        'name'          => $self->company_collection_name,
        'num_documents' => 0,
        'fields'        => [
            {
                'name'  => 'company_name',
                'type'  => 'string',
                'facet' => 0,
            },
            {
                'name'  => 'num_employees',
                'type'  => 'int32',
                'facet' => 0,
            },
            {
                'name'  => 'country',
                'type'  => 'string',
                'facet' => 1,
            }
        ],
        'default_sorting_field' => 'num_employees'
    };
}

sub DEMOLISH {
    my $typesense = $_[0]->typesense;
    $typesense->collections->delete_all if $typesense;
}

1;


t/lib/Test/Search/Typesense/Cached.pm
package Test::Search::Typesense::Cached;

use Moo;
use Test::Most ();
use Mojo::File;
use Digest::MD5 qw(md5_hex);
use Mojo::UserAgent::Mockable;
use FindBin;
use Search::Typesense::Types qw(
  Bool
  InstanceOf
);

extends 'Search::Typesense';

has '+_ua' => (
    is  => 'rw',
    isa => InstanceOf ['Mojo::UserAgent'],
);

has _force_cache => (
    is      => 'rw',
    isa     => Bool,
    default => 0,
);

sub assert_is_running {
    my $self = shift;
    my $is_running = eval { $self->next::method; 1 };
    unless ($is_running) {
        Test::Most::diag("Typesense is not running. Falling back to cache, if possible.");
        $self->_force_cache(1);
    }
}

sub BUILD {
    my $self = shift;
    my @tdir = ( $FindBin::Bin, '..', 't' );
    my @host = (
        $self->_make_slug( $self->host ),
        $self->_make_slug( $self->port )
    );

    # where we cache our Typesense results
    my $cache = Mojo::File->new( @tdir, 'cache', 'data', @host )->make_path;

    # checksums of our test files. If the md5 sum of the file contents is
    # different, we know we can't rely on the cached results
    my $checksum
      = Mojo::File->new( @tdir, 'cache', 'checksums', @host )->make_path;

    my $test_program = Mojo::File->new($0);
    $cache    = $cache->child( $test_program->basename('.t') );
    $checksum = $checksum->child( $test_program->basename('.t') );

    # checksum of the actual test file
    my $current_checksum = md5_hex( $test_program->slurp );
    my $ua   = $self->_get_user_agent( $cache, $checksum, $current_checksum );
    my $mode = $ua->can('mode') ? $ua->mode : '';

    my $url = $self->_url( [] );
    if ( 'record' eq $mode ) {
        $checksum->spurt($current_checksum);
        Test::Most::explain(
            "\nRecording all traffic to and from $url. This will be cached in $cache.\n\n"
        );
    }
    elsif ( 'playback' eq $mode ) {
        Test::Most::explain(
            "\nPlaying back cached traffic to and from $url from $cache.\n\n"
        );
    }
    else {
        Test::Most::explain(
            "\nWe are neither recording traffic nor playing it back.\n\n");
    }
    if ( my $app = $ua->server->app ) {

        # if we don't include this, we get tons of extra log lines spit out to
        # STDERR when running tests. Note we test for the existence of the app
        # brecause record mode doesn't use an app
        $app->log->level('fatal');
    }

    my $key = $self->api_key;
    $ua->on(
        start => sub {
            my ( $ua, $tx ) = @_;
            $tx->req->headers->header( 'Content-Type' => 'application/json' )
              ->header( 'X-TYPESENSE-API-KEY' => $key );
        }
    );
    $self->_ua($ua);
}

sub _get_user_agent {
    my ( $self, $cache, $checksum, $current_checksum ) = @_;

    # if we have PERL_TEST_TYPESENSE_MODE, assume that's the correct behavior
    # for Mojo::UserAgent, regardless of cache existence.
    local $ENV{PERL_TEST_TYPESENSE_MODE} if $self->_force_cache;

    my $mode = 'record';
    if ( defined( my $override_mode = $ENV{PERL_TEST_TYPESENSE_MODE} ) ) {
        if ( 'devel' eq $override_mode ) {
            return Mojo::UserAgent->new;
        }
        elsif ( 'record' eq $override_mode || 'playback' eq $override_mode ) {
            $mode = $override_mode;
        }
        else {
            croak(
                "PERL_TEST_TYPESENSE_MODE set to unknown value: '$override_mode'"
            );
        }
    }

    # otherwise, we'll default to 'record' for the useragent mode, unless we
    # have a cache and the test checksum is unchanged. In that case, we'll
    # assume the mode is 'playback'
    elsif ( -e $cache && -s _ ) {

        # The cache exists and is not empty. We assume it's good (famous last
        # words)
        if ( -e $checksum ) {
            my $cached_checksum = $checksum->slurp;
            if ( $current_checksum eq $cached_checksum ) {
                $mode = 'playback';
            }
            elsif ( $self->_force_cache ) {
                Test::Most::diag(
                    "Could not playback because cached checksum of $cached_checksum did not match new checksum of $current_checksum"
                );
            }
        }
        elsif ( $self->_force_cache ) {
            Test::Most::diag(
                "Could not playback because cached checksum file $checksum does not exist"
            );
        }
    }

    return Mojo::UserAgent::Mockable->new( mode => $mode, file => $cache );
}

sub _make_slug {
    my ( $self, $name ) = @_;
    $name = lc($name);
    $name =~ s/^\s+|\s+$//g;
    $name =~ s/\s+/_/g;
    $name =~ tr/-/_/;
    $name =~ s/__*/_/g;
    $name =~ s/\W//g;
    $name =~ tr/_/-/;
    $name =~ s/--/-/g;
    return $name;
}

1;
__END__

=head1 NAME

Test::Search::Typesense::Cached - Cache Typesense responses per test file

=head1 SYNOPSIS

    my $test       = Test::Search::Typesense->new;
    my $typesense  = $test->typesense;  # this is cached
    $typesense->collection->delete_all;

You don't instantiate this directly. Instead, you call
C<< Test::Search::Typesense->new >> and use the C<< ->typesense >> method
to fetch an instance of this module.

=head1 BEHAVIOR

This will behave just like L<Search::Typesense>, except that when you run
tests, you I<probably> don't need a Typesense server running any more.
Instead, the tests will playback responses cached in C<t/cache/data>. You
shouldn't have to worry about this.

If you're developing, you probably want to set the C<PERL_TEST_TYPESENSE_MODE>
environment variable to C<devel>:

    PERL_TEST_TYPESENSE_MODE=devel prove -lv t

That will stop all caching or playback of cached output.

To force recording output, use C<record>:

    PERL_TEST_TYPESENSE_MODE=record prove -lv t

To force reading playback (probably not a great idea):

    PERL_TEST_TYPESENSE_MODE=playback prove -lv t


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;

use 5.016000;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Perl interface to Typesense search engine.",
  "AUTHOR" => "Curtis \"Ovid\" Poe <ovid\@allaroundtheworld.fr>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Search-Typesense",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.016000",
  "NAME" => "Search::Typesense",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Mojolicious" => "9.11",
    "Moo" => "2.005_000",
    "Type::Tiny" => "1.011_011"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => 0,
    "Test::Most" => "0.37"
  },
  "VERSION" => "0.05",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Mojolicious" => "9.11",
  "Moo" => "2.005_000",
  "Test::More" => 0,
  "Test::Most" => "0.37",
  "Type::Tiny" => "1.011_011"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Search/Typesense/Types.pm
package Search::Typesense::Types;
# ABSTRACT: Types and type tools for Search::Typesense


use strict;
use warnings;
use Type::Library -base;
use Type::Utils -all;

# this gets us compile and compile_named
use Type::Params;

our $VERSION = '0.09';

our @EXPORT_OK;

BEGIN {
    extends qw(
      Types::Standard
      Types::Common::Numeric
      Types::Common::String
    );
    push @EXPORT_OK => (
        'compile',          # from Type::Params
        'compile_named',    # from Type::Params
    );
}

1;

__END__

=head1 SYNOPSIS

    package Search::Typesense;

    use Search::Typesense::Types qw(
      ArrayRef
      Dict
      Enum
      HashRef
      InstanceOf
      Str
      compile
    );

=head1 DESCRIPTION

This is an internal package for L<Search::Typesense>. It's probably overkill,
but if we want to be more strict later, this gives us the basics.

=head1 TYPE LIBRARIES

We automatically include the types from the following:

=over

=item * L<Types::Standard>

=item * L<Types::Common::Numeric>

=item * L<Types::Common::String>

=back

=head1 EXTRAS

The following extra functions are exported on demand or if use the C<:all> export tag.

=over

=item * C<compile>

See L<Type::Params>

=item * C<compile_named>

See L<Type::Params>

=item * C<slurpy>

See L<Types::Standard>

=back


lib/Search/Typesense/Version.pm
package Search::Typesense::Version;

# ABSTRACT: Version object for the Typesense server version

use Moo;
use Carp qw(croak);
use Search::Typesense::Types qw(
  NonEmptyStr
  PositiveOrZeroInt
);

our $VERSION = '0.09';

has version_string => (
    is       => 'ro',
    isa      => NonEmptyStr,
    required => 1,
);

has [qw/major minor patch/] => (
    is       => 'rwp',
    isa      => PositiveOrZeroInt,
    init_arg => undef,
);

sub BUILD {
    my $self    = shift;
    my $version = $self->version_string;

    # this isn't quite semver, but it seems to fit what Typesense is doing
    # See https://semver.org/ if this breaks
    unless ( $version =~ /^\d+\.\d+\.\d+$/a ) {
        croak("Invalid version string: $version");
    }
    my @version = split /\./ => $version;
    $self->_set_major( $version[0] );
    $self->_set_minor( $version[1] );
    $self->_set_patch( $version[2] );
}

sub comparator {
    my $self = shift;
    return sprintf "%03d%03d%03d" => $self->major, $self->minor, $self->patch;
}

1;

__END__

=head1 DESCRIPTION

Do not use directly. This is returned by the C<typesense_version> method from
L<Search::Typesense>.

=head1 METHODS

=head2 C<version_string>

    my $version        = $typesense->typesense_version;
    my $version_string = $version->version_string;

Returns the semantic version string, such as C<0.19.0>.

=head2 C<major>

    my $version = $typesense->typesense_version;
    my $major   = $version->major;

Returns the major version number.

=head2 C<minor>

    my $version = $typesense->typesense_version;
    my $minor   = $version->minor;

Returns the minor version number.

=head2 C<patch>

    my $version = $typesense->typesense_version;
    my $patch   = $version->patch;

Returns the patch version number.

=head2 C<comparator>

Returns a numeric string suitable for numeric comparisons between versions.


lib/Search/Typesense/Document.pm
package Search::Typesense::Document;
# ABSTRACT: CRUD for Typesense documents

use v5.16.0;

use Moo;
with qw(Search::Typesense::Role::Request);

use Mojo::JSON qw(decode_json encode_json);
use Search::Typesense::Types qw(
  ArrayRef
  Enum
  HashRef
  InstanceOf
  NonEmptyStr
  Str
  compile
);

=head1 SYNOPSIS

    my $typesense = Search::Typesense->new(
        host    => $host,
        api_key => $key,
    );
    my $documents = $typesense->documents;

The instantiation of this module is for internal use only. The methods are
public.

=cut

our $VERSION = '0.09';

=head2 C<create>

    my $document = $typesense->documents->create($collection, \%data);

Arguments and response as shown at L<https://typesense.org/docs/0.19.0/api/#index-document>

=cut

sub create {
    my ( $self, $collection, $document ) = @_;
    state $check = compile( NonEmptyStr, HashRef );
    ( $collection, $document ) = $check->( $collection, $document );
    return $self->_POST(
        path => [ 'collections', $collection, 'documents' ],
        body => $document
    );
}

=head2 C<upsert>

    my $document = $typesense->documents->upsert($collection, \%data);

Arguments and response as shown at L<https://typesense.org/docs/0.19.0/api/#upsert>

=cut

sub upsert {
    my ( $self, $collection, $document ) = @_;
    state $check = compile( NonEmptyStr, HashRef );
    ( $collection, $document ) = $check->( $collection, $document );

    return $self->_POST(
        path  => [ 'collections', $collection, 'documents' ],
        body  => $document,
        query => { action => 'upsert' },
    );
}

=head2 C<update>

    my $document = $typesense->documents->update($collection, $document_id, \%data);

Arguments and response as shown at L<https://typesense.org/docs/0.19.0/api/#update-document>

=cut

sub update {
    my ( $self, $collection, $document_id, $updates ) = @_;
    state $check = compile( NonEmptyStr, NonEmptyStr, HashRef );
    ( $collection, $document_id, $updates )
      = $check->( $collection, $document_id, $updates );
    return $self->_PATCH(
        path => [ 'collections', $collection, 'documents', $document_id ],
        body => $updates
    );
}

=head2 C<delete>

    my $document = $typesense->documents->delete($collection_name, $document_id);

Arguments and response as shown at L<https://typesense.org/docs/0.19.0/api/#delete-document>

=cut

sub delete {
    my ( $self, $collection, $document_id ) = @_;
    state $check = compile( NonEmptyStr, NonEmptyStr );
    ( $collection, $document_id ) = $check->( $collection, $document_id );
    return $self->_DELETE(
        path => [ 'collections', $collection, 'documents', $document_id ] );
}

=head2 C<export>

    my $export = $typesense->documents->export($collection_name);

Response as shown at L<https://typesense.org/docs/0.19.0/api/#export-documents>

(An arrayref of hashrefs)

=cut

sub export {
    my ( $self, $collection ) = @_;
    state $check = compile(NonEmptyStr);
    ($collection) = $check->($collection);
    my $tx = $self->_GET(
        path => [ 'collections', $collection, 'documents', 'export' ],
        return_transaction => 1
    ) or return;    # 404
    return [ map { decode_json($_) } split /\n/ => $tx->res->body ];
}

=head2 C<import>

    my $response = $typesense->documents->import(
      $collection_name,
      $action,
      \@documents,
   );

Response as shown at L<https://typesense.org/docs/0.19.0/api/#import-documents>

C<$action> must be one of C<create>, C<update>, or C<upsert>.

=cut

sub import {
    my $self = shift;
    state $check = compile(
        NonEmptyStr,
        Enum [qw/create upsert update/],
        ArrayRef [HashRef],
    );
    my ( $collection, $action, $documents ) = $check->(@_);
    my $body = join "\n" => map { encode_json($_) } @$documents;

    my $tx = $self->_POST(
        path  => [ 'collections', $collection, 'documents', "import" ],
        body  => $body,
        query => { action => $action },
        return_transaction => 1,
    );
    my $response = $tx->res->json;
    if ( exists $response->{success} ) {
        $response->{success} += 0;
    }
    return $response;
}

1;


lib/Search/Typesense/Collection.pm
package Search::Typesense::Collection;

# ABSTRACT: Collection - CRUD for Typesense collections

use v5.16.0;

use Moo;
with qw(Search::Typesense::Role::Request);

use Carp 'croak';
use Search::Typesense::Types qw(
  HashRef
  InstanceOf
  NonEmptyStr
  Str
  compile
);

=head1 SYNOPSIS

    my $typesense = Search::Typesense->new(
        host    => $host,
        api_key => $key,
    );
    my $collections = $typesense->collections;

The instantiation of this module is for internal use only. The methods are
public.

=head2 C<get>

    if ( my $collections = $typesense->collections->get ) {
        # returns all collections
    }
    if ( my $collections = $typesense->collections->get($collection_name) ) {
        # returns collection matching $collection_name, if any
    }

Response shown at L<https://typesense.org/docs/0.19.0/api/#retrieve-collection>

=cut

our $VERSION = '0.09';

=head2 C<search>

    my $results = $typesense->collections->search($collection_name, {q => 'London'});

The parameters for C<$query> are defined at
L<https://typesense.org/docs/0.19.0/api/#search-collection>, as are the results.

Unlike other methods, if we find nothing, we still return the data structure
(instead of C<undef> instead of a 404 exception).

=cut

sub search {
    my ( $self, $collection, $query ) = @_;
    state $check = compile( NonEmptyStr, HashRef );
    ( $collection, $query ) = $check->( $collection, $query );

    unless ( exists $query->{q} ) {
        croak("Query parameter 'q' is required for searching");
    }
    my $tx = $self->_GET(
        path  => [ 'collections', $collection, 'documents', 'search' ],
        query => $query,
        return_transaction => 1,
    ) or return;
    return $tx->res->json;
}

sub get {
    my ( $self, $collection ) = @_;
    state $check = compile(Str);
    my @collection = $check->( $collection // '' );
    return $self->_GET( path => [ 'collections', @collection ] );
}

=head2 C<create>

    my $collection = $typesense->collections->create(\%definition);

Arguments and response as shown at
L<https://typesense.org/docs/0.19.0/api/#create-collection>

=cut

sub create {
    my ( $self, $collection_definition ) = @_;
    state $check = compile(HashRef);
    ($collection_definition) = $check->($collection_definition);
    my $fields = $collection_definition->{fields};

    foreach my $field (@$fields) {
        if ( exists $field->{facet} ) {
            $field->{facet}
              = $field->{facet} ? Mojo::JSON->true : Mojo::JSON->false;
        }
    }

    return $self->_POST(
        path => ['collections'],
        body => $collection_definition
    );
}

=head2 C<delete>

    my $response = $typesense->collections->delete($collection_name);

Response shown at L<https://typesense.org/docs/0.19.0/api/#drop-collection>

=cut

sub delete {
    my ( $self, $collection ) = @_;
    state $check = compile(NonEmptyStr);
    ($collection) = $check->($collection);
    return $self->_DELETE( path => [ 'collections', $collection ] );
}

=head2 C<delete_all>

    $typesense->collections->delete_all;

Deletes everything from Typesense. B<Use with caution>!

=cut

sub delete_all {
    my ($self) = @_;
    my $collections = $self->get;
    foreach my $collection (@$collections) {
        my $name = $collection->{name};
        $self->delete($name);
    }
}

1;


lib/Search/Typesense.pm
package Search::Typesense;
# ABSTRACT: Perl interface to the Typesense search engine

use v5.16.0;

use Moo;
with 'Search::Typesense::Role::Request';

use Mojo::JSON qw(decode_json encode_json);
use Mojo::UserAgent;
use Mojo::URL;
use Carp qw(croak);

use Search::Typesense::Document ();
use Search::Typesense::Collection;
use Search::Typesense::Version;
use Search::Typesense::Types qw(
  ArrayRef
  Bool
  Enum
  HashRef
  InstanceOf
  NonEmptyStr
  PositiveInt
  compile
);

=head1 SYNOPSIS

    my $typesense = Search::Typesense->new(
        host      => $host,    # required
        api_key   => $key,     # required
        port      => $port,    # defaults to 8108
        use_https => $bool,    # defaults to true
    );
    
    my $results = $typesense->search(
        $collection_name,
        { q => 'Search String' },
    );
    if ( $results->{found} ) {
        foreach my $hit ( @{ $results->{hits} } ) {
            ...;
        }
    }

L<Check here for a comparison to ElasticSearch and similar technologies|https://typesense.org/typesense-vs-algolia-vs-elasticsearch-vs-meilisearch/>.

=head1 DESCRIPTION

B<ALPHA CODE>. The interface can and will change without warning.

This is an interface to the L<Typesense|https://typesense.org/> search
engine. Most methods will do one of three things:

=over 4

=item * Return results as defined in the Typesense documentation (listed per section)

=item * Return nothing if Typesense returns a 404.

=item * C<croak> if Typesense returns an error.

=back

=cut

our $VERSION = '0.09';

has collections => (
    is       => 'lazy',
    isa      => InstanceOf ['Search::Typesense::Collection'],
    init_arg => undef,
    handles  => [qw/search/],
    builder  => sub {
        my $self = shift;
        return Search::Typesense::Collection->new(
            user_agent => $self->_ua,
            url        => $self->_url_base,
        );
    },
);

has documents => (
    is       => 'lazy',
    isa      => InstanceOf ['Search::Typesense::Document'],
    init_arg => undef,
    builder  => sub {
        my $self = shift;
        return Search::Typesense::Document->new(
            user_agent => $self->_ua,
            url        => $self->_url_base,
        );
    },
);

# this sub without a body is called a "forward declaration" and it allows the
# requires() in Search::Typesense::Role::Request to realize that we really do
# provide these methods. In this case, it also prevents the following error
# from being generated by the roles consumed:
#
#  Error:  You cannot overwrite a locally defined method (_ua) with a reader
#  at .../lib/Search/Typesense.pm line 121.

sub _ua;
has _ua => (
    is      => 'lazy',
    isa     => InstanceOf ['Mojo::UserAgent'],
    builder => sub {
        my $self = shift;
        my $ua   = Mojo::UserAgent->new;
        my $key  = $self->api_key;
        $ua->on(
            start => sub {
                my ( $ua, $tx ) = @_;
                $tx->req->headers->header(
                    'Content-Type' => 'application/json' )
                  ->header( 'X-TYPESENSE-API-KEY' => $key );
            }
        );
        return $ua;
    },
);

sub _url_base;
has _url_base => (
    is      => 'lazy',
    isa     => InstanceOf ['Mojo::URL'],
    builder => sub {
        my $self = shift;
        my $url  = Mojo::URL->new;
        $url->scheme( $self->use_https ? 'https' : 'http' );
        $url->host( $self->host );
        $url->port( $self->port );
        return $url;
    },
);

has use_https => (
    is       => 'ro',
    isa      => Bool,
    required => 1,
    default  => 1,
);

has api_key => (
    is       => 'ro',
    isa      => NonEmptyStr,
    required => 1,
);

has host => (
    is       => 'ro',
    isa      => NonEmptyStr,
    required => 1,
);

has port => (
    is      => 'ro',
    isa     => PositiveInt,
    default => 8108,
);

sub BUILD {
    my $self = shift;
    $self->assert_is_running;
}

=head1 CONSTRUCTOR

The constructor takes a list (or hashref) of key/value pairs.

    my $typesense = Search::Typesense->new(
        host      => $host,    # required
        api_key   => $key,     # required
        port      => $port,    # defaults to 8108
        use_https => $bool,    # defaults to true
    );

=head2 C<api_key>

The api key to which will be sent as the C<X-TYPESENSE-API-KEY> header.

=head2 C<host>

The hostname to connect to.

=head2 C<port>

Optional port number to connect to. Defaults to 8108 if not supplied.

=head2 C<use_https>

Optional boolean. Whether or not to connect to Typesense over https. Default true.

=head1 METHODS

For CRUD operations on collections and documents, see the documentation for
C<collections> (L<Search::Typesense::Collection>) and C<documents>
(L<Search::Typesense::Document>).

=head2 C<collections>

    my $collections = $typesense->collections;
    my $collection  = $collections->get($collection_name);
    my $results     = $collections->search($collection_name, {q => 'London'});

Returns an instance of L<Search::Typesense::Collection> for managing Typesense collections.

=head2 C<search>

    my $results = $typesense->search($collection_name, {q => 'London'});

Shorthand that delegated to C<< $typesense->collections->search(...) >>.

We provide this on the top-level C<$typesense> object because this is the
common case.

=head2 C<documents>

    my $documents = $typesense->documents;
    my $document  = $documents->delete($collection_name, $document_id);

Returns an instance of L<Search::Typesense::Document> for managing Typesense documents.

=head2 C<assert_is_running>

    $typesense->assert_is_running;

This does nothing if we can connect to Typesense. Otherwise, this method will
C<croak> with a message explaining the error.

=cut

sub assert_is_running {
    my $self = shift;
    $self->_GET( path => ['health'] );
}

=head2 C<typesense_version>

    my $version = $typesense->typesense_version;

Returns an instance of L<Search::Typesense::Version>.

If your version of Typesense is older than C<0.8.0>, this method will return
nothing.

=cut

sub typesense_version {
    my $self   = shift;
    my $result = $self->_GET( path => ['debug'] ) or return;
    return Search::Typesense::Version->new(
        version_string => $result->{version} );
}

1;

__END__

=head1 INTERNATIONALIZATION (I18N)

Currently Typesense supports languages that use spaces as a word separator. In
the future, a new tokenizer will be added to support languages such as Chinese
or Japanese. I do not know the timeframe for this.

=head1 AUTHOR

Curtis "Ovid" Poe, C<< <ovid at allaroundtheworld.fr> >>

=head1 BUGS

Please report any bugs or feature requests to
C<https://github.com/Ovid/Search-Typesense/issues>.  I will be notified, and
then you'll automatically be notified of progress on your bug as I make
changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Search::Typesense

You can also look for information at:

=over 4

=item * Github Repo

L<https://github.com/Ovid/Search-Typesense/>

=item * Issue Tracker

L<https://github.com/Ovid/Search-Typesense/issues>

=item * Search CPAN

L<https://metacpan.org/release/Search-Typesense>

=back

=head1 ACKNOWLEDGEMENTS

Thanks for Sebastian Reidel and Matt Trout for feedback.

=head1 LICENSE AND COPYRIGHT

This software is Copyright (c) 2021 by Curtis "Ovid" Poe.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)
